### DISCUSSION TITLE
Remove `cmake: false` in C target (and use CMake to finally create a nice, modular build :sparkles:)?

### DISCUSSION THREAD

**Discussion by petervdonovan** [lf-lang/lingua-franca#1245] [Ideas]
This is a continuation of the discussion started [here](https://github.com/lf-lang/lingua-franca/pull/1207#discussion_r901878295). The proposal is to remove the `cmake` target property and to remove the logic in the C Generator that picks out which files to copy into the `src-gen` folder so that that logic can be implemented in CMake instead.

Code that could be removed from our compiler includes:
* [CCompiler.java](https://github.com/lf-lang/lingua-franca/blob/master/org.lflang/src/org/lflang/generator/c/CCompiler.java)
* [CCoreFilesUtils.java](https://github.com/lf-lang/lingua-franca/blob/master/org.lflang/src/org/lflang/generator/c/CCoreFilesUtils.java)
* A few lines from [TargetProperty.java](https://github.com/lf-lang/lingua-franca/blob/9cb9ae782940721a771793e44153a3fc85d4d321/org.lflang/src/org/lflang/TargetProperty.java#L1317)
* Parts of `doGenerate` in `CGenerator.java` around [here](https://github.com/lf-lang/lingua-franca/blob/9cb9ae782940721a771793e44153a3fc85d4d321/org.lflang/src/org/lflang/generator/c/CGenerator.java#L556), [here](https://github.com/lf-lang/lingua-franca/blob/9cb9ae782940721a771793e44153a3fc85d4d321/org.lflang/src/org/lflang/generator/c/CGenerator.java#L891), and [here](https://github.com/lf-lang/lingua-franca/blob/9cb9ae782940721a771793e44153a3fc85d4d321/org.lflang/src/org/lflang/generator/c/CGenerator.java#L1081)
* Tests, including [this](https://github.com/lf-lang/lingua-franca/blob/master/test/C/src/target/HelloWorldCCPPNoCMake.lf), [this](https://github.com/lf-lang/lingua-franca/blob/master/test/C/src/target/HelloWorldFederatedCCPPNoCMake.lf), [this](https://github.com/lf-lang/lingua-franca/blob/master/test/C/src/target/HelloWorldFederatedNoCMake.lf), and [this](https://github.com/lf-lang/lingua-franca/blob/master/test/C/src/target/HelloWorldNoCMake.lf).
* Documentation, including [this](https://www.lf-lang.org/docs/handbook/target-declaration#cmake), [this](https://www.lf-lang.org/docs/handbook/target-declaration#cmake), and [this](https://www.lf-lang.org/docs/handbook/target-declaration#cmake), as well as [this user warning](https://github.com/lf-lang/lingua-franca/blob/a013ba8c39191aab3636577bc46516a674b19694/org.lflang/src/org/lflang/generator/c/CCmakeGenerator.java#L223)

**The total amount of code deleted might be only a few hundred lines, and that would be either partially or completely offset by the rewriting of the deleted code in CMake.** However, I think complexity would be reduced in terms of:
* the amount of redundancy in the code we must maintain
* the number of complex tasks that must be handled in the C generator (not enough separation of concerns)
* the number of files (in distinct parts of the code base!) that must be touched in order to keep the main `lingua-franca` repo in sync with a change to the file structure of the C runtime
* the amount of explanation required to bring new users up to speed

**Redundancy and separation of concerns are the primary motivators** of this suggestion. An example of redundancy is that `reactor-c` already has a skeletal toy implementation of a CMake-only build process because we needed it for unit tests.

This issue is not very important to me. The reason why I bring it up is that after discussing it with Marten and Soroush, I am still not aware of use-cases for `cmake: false`, and if anyone knows of use-cases, I would like to understand them. To my knowledge, CMake probably can be used for cross-compilation to baremetal/embedded platforms; furthermore, if it cannot be used for that, then a drop-in replacement for CMake might be supplied using the [build](https://github.com/lf-lang/lingua-franca/blob/a013ba8c39191aab3636577bc46516a674b19694/org.lflang/src/org/lflang/generator/c/CCmakeGenerator.java#L223) target property.

**Comment by petervdonovan** [comment]
[This](https://github.com/lf-lang/lingua-franca/blob/2011ccacb472451651d7e0d391127c8060fc9e54/org.lflang/src/org/lflang/generator/python/PythonGenerator.java#L645) seems like a roadblock that I did not know about:
>  targetConfig.useCmake = false; // Force disable the CMake because
        // it interferes with the Python target functionality

**Reply by Soroosh129** [reply]
Well, I forgot how CMake actually interferes here. Wouldn't setting `targetConfig.noCompile = true` be enough?

**Reply by petervdonovan** [reply]
> Wouldn't setting `targetConfig.noCompile = true` be enough?

I suppose I will have to find out.

Hi @Soroosh129, I have a similar question -- It looks like the bash script in `FedLauncher` uses a compile command for remote federates. In order to move forward with this, would I need to update the bash script so that it invokes CMake?

This just came up because the way the C generator manages files is an obstacle for changes in the file structure of the runtime...

**Reply by Soroosh129** [reply]
> In order to move forward with this, would I need to update the bash script so that it invokes CMake?

Yes, I think that has to be updated.
